src/graphql/index.js

```diff
@@ -37,29 +38,52 @@ const EMPTY_CONTEXT = {
  * @returns {object}
  */
 export async function getContext({ ctx: { req } }) {
-  const jwt = (req.headers.authorization || '').replace(/^Bearer /, '');
-  if (!jwt || !verify(jwt)) {
-    return EMPTY_CONTEXT;
-  }
+  const authorization = req.headers.authorization || '';
+  if (authorization.toLowerCase().startsWith('line ')) {
+    const idToken = authorization.replace(/^line /i, '');
+    const parsed = await verifyIDToken(idToken);
 
-  const parsed = read(jwt);
+    if (!parsed || !parsed.sub) {
+      return EMPTY_CONTEXT;
+    }
 
-  if (!parsed || !parsed.sub) {
-    return EMPTY_CONTEXT;
-  }
+    // The user may have no context at all
+    // (Brand-new user activating rich mebnu)
+    // At least give empty context
+    //
+    const context = (await redis.get(parsed.sub)) || {};
+
+    return {
+      userId: parsed.sub,
+      userContext: context,
+    };
+  } else if (authorization.toLowerCase().startsWith('bearer ')) {
+    const jwt = (req.headers.authorization || '').replace(/^Bearer /i, '');
+    if (!jwt || !verify(jwt)) {
+      return EMPTY_CONTEXT;
+    }
+
+    const parsed = read(jwt);
 
-  const context = await redis.get(parsed.sub);
+    if (!parsed || !parsed.sub) {
+      return EMPTY_CONTEXT;
+    }
+
+    const context = await redis.get(parsed.sub);
+
+    return {
+      userId: parsed.sub,
+      userContext:
+        context &&
+        context.data &&
+        parsed.sessionId &&
+        context.data.sessionId === parsed.sessionId
+          ? context
+          : null,
+    };
+  }
 
-  return {
-    userId: parsed.sub,
-    userContext:
-      context &&
-      context.data &&
-      parsed.sessionId &&
-      context.data.sessionId === parsed.sessionId
-        ? context
-        : null,
-  };
+  throw new Error('Invalid authentication type.');
```

Comment:
If we allow anonymous user and reject unsupported authentication method or just return `EMPTY_CONTEXT` here?

Reply:
I think we should return `EMPTY_CONTEXT` here.

`getContext` is invoked by all GraphQL calls, providing `context` in resolvers. The GraphQL APIs that don't need authentication also calls `getContext`.

We should only throw exception within `auth` directive (`directives/auth.js`), because that directive is the one that is responsible for guarding authenticating endpoints. `getContext` only provides GraphQL contexts, as its name suggests.